b2sdk/utils/thread_pool.py

```diff
@@ -9,27 +9,84 @@
 ######################################################################
 from __future__ import annotations
 
-from concurrent.futures import ThreadPoolExecutor
+from concurrent.futures import Future, ThreadPoolExecutor
+from typing import Callable
 
 from b2sdk.utils import B2TraceMetaAbstract
 
 
+class LazyThreadPool:
+    """
+    Lazily initialized thread pool.
+    """
+
+    _THREAD_POOL_FACTORY = ThreadPoolExecutor
+
+    def __init__(self, max_workers: int | None = None, **kwargs):
+        self._max_workers = max_workers
+        self._thread_pool: ThreadPoolExecutor | None = None
+        super().__init__(**kwargs)
+
+    def submit(self, fn: Callable, *args, **kwargs) -> Future:
+        if self._thread_pool is None:
+            self._thread_pool = self._THREAD_POOL_FACTORY(self._max_workers)
+        return self._thread_pool.submit(fn, *args, **kwargs)
+
+    def set_size(self, max_workers: int) -> None:
+        """
+        Set the size of the thread pool.
+
+        This operation will block until all tasks in the current thread pool are completed.
+
+        :param max_workers: New size of the thread pool
+        :return: None
+        """
+        if self._max_workers == max_workers:
+            return
+        old_thread_pool = self._thread_pool
+        self._thread_pool = self._THREAD_POOL_FACTORY(max_workers=max_workers)
+        if old_thread_pool is not None:
+            old_thread_pool.shutdown(wait=True)
+        self._max_workers = max_workers
+
+    def get_size(self) -> int | None:
+        return self._max_workers
+
+
 class ThreadPoolMixin(metaclass=B2TraceMetaAbstract):
     """
     Mixin class with ThreadPoolExecutor.
     """
-    DEFAULT_THREAD_POOL_CLASS = staticmethod(ThreadPoolExecutor)
+
+    DEFAULT_THREAD_POOL_CLASS = LazyThreadPool
 
     def __init__(
         self,
         thread_pool: ThreadPoolExecutor | None = None,
         max_workers: int | None = None,
-        **kwargs
+        **kwargs,
     ):
         """
         :param thread_pool: thread pool to be used
         :param max_workers: maximum number of worker threads (ignored if thread_pool is not None)
         """
-        self._thread_pool = thread_pool if thread_pool is not None \
-            else self.DEFAULT_THREAD_POOL_CLASS(max_workers=max_workers)
+        self._thread_pool = (
+            thread_pool
+            if thread_pool is not None else self.DEFAULT_THREAD_POOL_CLASS(max_workers=max_workers)
+        )
```

Comment:
I think type discrepancy appeared here. The `__init__` method says:
```python
thread_pool: ThreadPoolExecutor | None = None,
```
and by default `LazyThreadPool` is assigned. But `LazyThreadPool` is not a subclass of `ThreadPoolExecutor` (it doesn't inherit from `ThreadPoolExecutor` class, and thus doesn't provide its interface, for example, the `map()` method).

Reply:
hm, actually, this is ok

i.e. this typing is for the argument of constructor and not instance attribute, so yea, initializer says it accepts only instances of `ThreadPoolExecutor`, but in fact could also accept some other things

this is the same as saying
```
def f(n: int | None = None) -> int | float:
  return n or 1.5
```
It would be different if the annotation was on `self._thread_pool` itself and not just the initializer argument

Anyhow; for clarity I will replace this with https://typing-extensions.readthedocs.io/en/latest/index.html#Protocol